src/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/PerformanceCounterLib.cs

```diff
@@ -638,83 +642,87 @@ internal bool FindCustomCategory(string category, out PerformanceCounterCategory
             RegistryKey baseKey = null;
             categoryType = PerformanceCounterCategoryType.Unknown;
 
-            if (_customCategoryTable == null)
+            //race with CloseTables
```

Comment:
This is a lot of code to put under a lock; this is not only synchronizing with CloseTables, but with other calls to FindCustomCategory.  Might that be a performance/scalability problem?

I would think a better way to deal with this would be to instead just change this method to not refer to _customCategoryTable other than for the initial checking/setting of the field, e.g.
```C#
Hashtable table =
    _customCategoryTable ??
    Interlocked.CompareExchange(ref _customCategoryTable, new Hashtable(StringComparer.OrdinalIgnoreCase), null) ??
    _customCategoryTable;
```
and then use `table` for the rest of the method rather than using `_customCategoryTable`.

Reply:
>Might that be a performance/scalability problem?

i followed internal FindCustomCategory() and is used only in init path today, no hot path.
However your solution is better, so i agree with you. Thank's a lot.